File: project-management-backend/app/services/admin_service.py
"""
Admin Service
Lógica de negocio para el panel de administración (SUPERADMIN)
"""

from app import db
from app.models import User, Project, Task, Membership, AuditLog
from sqlalchemy import func, desc
from datetime import datetime, timedelta


class AdminService:
    """Servicio para operaciones de administración"""
    
    @staticmethod
    def get_all_users(page=1, per_page=20, search=None, status=None):
        """
        Obtener todos los usuarios con paginación y filtros
        
        Args:
            page: Número de página
            per_page: Usuarios por página
            search: Búsqueda por nombre o email
            status: Filtrar por estado (active, inactive)
        
        Returns:
            dict: Usuarios paginados con metadata
        """
        query = User.query
        
        # Filtro de búsqueda
        if search:
            search_filter = f"%{search}%"
            query = query.filter(
                db.or_(
                    User.name.ilike(search_filter),
                    User.email.ilike(search_filter)
                )
            )
        
        # Filtro de estado
        if status:
            query = query.filter_by(status=status)
        
        # Ordenar por fecha de creación (más recientes primero)
        query = query.order_by(desc(User.created_at))
        
        # Paginación
        pagination = query.paginate(page=page, per_page=per_page, error_out=False)
        
        return {
            'users': pagination.items,
            'total': pagination.total,
            'page': page,
            'per_page': per_page,
            'total_pages': pagination.pages,
            'has_next': pagination.has_next,
            'has_prev': pagination.has_prev
        }
    
    @staticmethod
    def get_all_projects(page=1, per_page=20, search=None, status=None):
        """
        Obtener todos los proyectos con paginación y filtros
        
        Args:
            page: Número de página
            per_page: Proyectos por página
            search: Búsqueda por nombre
            status: Filtrar por estado (active, inactive)
        
        Returns:
            dict: Proyectos paginados con metadata
        """
        query = Project.query
        
        # Filtro de búsqueda
        if search:
            search_filter = f"%{search}%"
            query = query.filter(Project.name.ilike(search_filter))
        
        # Filtro de estado
        if status:
            query = query.filter_by(status=status)
        
        # Ordenar por fecha de creación (más recientes primero)
        query = query.order_by(desc(Project.created_at))
        
        # Paginación
        pagination = query.paginate(page=page, per_page=per_page, error_out=False)
        
        return {
            'projects': pagination.items,
            'total': pagination.total,
            'page': page,
            'per_page': per_page,
            'total_pages': pagination.pages,
            'has_next': pagination.has_next,
            'has_prev': pagination.has_prev
        }
    
    @staticmethod
    def get_audit_logs(page=1, per_page=50, user_id=None, action=None, days=7):
        """
        Obtener logs de auditoría con filtros
        
        Args:
            page: Número de página
            per_page: Logs por página
            user_id: Filtrar por usuario
            action: Filtrar por acción
            days: Últimos N días (default: 7)
        
        Returns:
            dict: Logs paginados con metadata
        """
        query = AuditLog.query
        
        # Filtro de fecha (últimos N días)
        if days:
            date_from = datetime.utcnow() - timedelta(days=days)
            query = query.filter(AuditLog.created_at >= date_from)
        
        # Filtro de usuario
        if user_id:
            query = query.filter_by(user_id=user_id)
        
        # Filtro de acción
        if action:
            query = query.filter(AuditLog.action.ilike(f"%{action}%"))
        
        # Ordenar por fecha (más recientes primero)
        query = query.order_by(desc(AuditLog.created_at))
        
        # Paginación
        pagination = query.paginate(page=page, per_page=per_page, error_out=False)
        
        return {
            'logs': pagination.items,
            'total': pagination.total,
            'page': page,
            'per_page': per_page,
            'total_pages': pagination.pages,
            'has_next': pagination.has_next,
            'has_prev': pagination.has_prev
        }
    
    @staticmethod
    def update_user_status(user_id, new_status):
        """
        Activar o desactivar un usuario
        
        Args:
            user_id: ID del usuario
            new_status: 'active' o 'disabled'
        
        Returns:
            User: Usuario actualizado
        """
        user = User.query.get_or_404(user_id)
        
        # No permitir desactivar SUPERADMIN
        if user.role == 'SUPERADMIN' and new_status == 'disabled':
            raise ValueError('No se puede desactivar un SUPERADMIN')
        
        user.status = new_status
        user.updated_at = datetime.utcnow()
        
        db.session.commit()
        
        return user
    
    @staticmethod
    def update_project_status(project_id, new_status):
        """
        Activar o desactivar un proyecto
        
        Args:
            project_id: ID del proyecto
            new_status: 'active' o 'disabled'
        
        Returns:
            Project: Proyecto actualizado
        """
        project = Project.query.get_or_404(project_id)
        
        project.status = new_status
        project.updated_at = datetime.utcnow()
        
        db.session.commit()
        
        return project
    
    @staticmethod
    def get_global_stats():
        """
        Obtener estadísticas globales de la plataforma
        
        Returns:
            dict: Estadísticas globales
        """
        # Contar usuarios
        total_users = User.query.count()
        active_users = User.query.filter_by(status='active').count()
        users_by_role = db.session.query(
            User.role,
            func.count(User.id)
        ).group_by(User.role).all()
        
        # Contar proyectos
        total_projects = Project.query.count()
        active_projects = Project.query.filter_by(status='active').count()
        
        # Contar tareas
        total_tasks = Task.query.count()
        tasks_by_status = db.session.query(
            Task.status,
            func.count(Task.id)
        ).group_by(Task.status).all()
        
        # Contar membresías
        total_memberships = Membership.query.count()
        active_memberships = Membership.query.filter_by(status='active').count()
        
        # Usuarios nuevos (últimos 30 días)
        date_30_days_ago = datetime.utcnow() - timedelta(days=30)
        new_users_30d = User.query.filter(
            User.created_at >= date_30_days_ago
        ).count()
        
        # Proyectos nuevos (últimos 30 días)
        new_projects_30d = Project.query.filter(
            Project.created_at >= date_30_days_ago
        ).count()
        
        # Tareas creadas (últimos 30 días)
        new_tasks_30d = Task.query.filter(
            Task.created_at >= date_30_days_ago
        ).count()
        
        return {
            'users': {
                'total': total_users,
                'active': active_users,
                'inactive': total_users - active_users,
                'by_role': {role: count for role, count in users_by_role},
                'new_last_30_days': new_users_30d
            },
            'projects': {
                'total': total_projects,
                'active': active_projects,
                'inactive': total_projects - active_projects,
                'new_last_30_days': new_projects_30d
            },
            'tasks': {
                'total': total_tasks,
                'by_status': {status: count for status, count in tasks_by_status},
                'new_last_30_days': new_tasks_30d
            },
            'memberships': {
                'total': total_memberships,
                'active': active_memberships,
                'inactive': total_memberships - active_memberships
            }
        }
    
    @staticmethod
    def get_health_check():
        """
        Health check del sistema
        
        Returns:
            dict: Estado del sistema
        """
        try:
            # Verificar conexión a la base de datos
            db.session.execute(db.text('SELECT 1'))
            db_status = 'healthy'
            db_message = 'Database connection OK'
        except Exception as e:
            db_status = 'unhealthy'
            db_message = f'Database error: {str(e)}'
        
        # Información del sistema
        return {
            'status': db_status,
            'timestamp': datetime.utcnow().isoformat(),
            'database': {
                'status': db_status,
                'message': db_message
            },
            'version': '1.0.0'
        }

File: project-management-frontend/services/adminService.ts
import { api } from '@/lib/api'
import { mapUserFromBackend, mapProjectFromBackend } from '@/lib/mappers'
import type { 
  UsersListResponse, 
  ProjectsListResponse, 
  AuditLogsResponse,
  GlobalStatsResponse,
  HealthCheckResponse,
  BackendAuditLog
} from '@/lib/api-types'
import type { User, Project, AuditLog, HealthCheck } from '@/mock/types'

// ============================================
// TIPOS DE PAGINACIÓN
// ============================================

export interface PaginatedResult<T> {
  items: T[]
  total: number
  page: number
  per_page: number
  total_pages: number
  has_next: boolean
  has_prev: boolean
}

// ============================================
// MAPEO DE AUDIT LOG
// ============================================

function mapAuditLogFromBackend(log: BackendAuditLog): AuditLog {
  return {
    id: log.id,
    action: log.action,
    user_id: log.user_id,
    user_email: log.user_email,
    project_id: log.project_id,
    details: log.details,
    ip: log.ip,
    created_at: log.created_at,
  }
}

// ============================================
// LIST ALL USERS (Listar todos los usuarios)
// ============================================

export async function fetchAllUsers(params?: {
  page?: number
  per_page?: number
  search?: string
  status?: string
}): Promise<PaginatedResult<User>> {
  try {
    const query = new URLSearchParams()
    if (params?.page) query.append('page', params.page.toString())
    if (params?.per_page) query.append('per_page', params.per_page.toString())
    if (params?.search) query.append('search', params.search)
    if (params?.status) query.append('status', params.status)
    
    const response = await api.get<UsersListResponse>(
      `/admin/users${query.toString() ? '?' + query.toString() : ''}`
    )
    
    return {
      items: response.users.map(mapUserFromBackend),
      total: response.pagination.total,
      page: response.pagination.page,
      per_page: response.pagination.per_page,
      total_pages: response.pagination.total_pages,
      has_next: response.pagination.has_next,
      has_prev: response.pagination.has_prev,
    }
  } catch (error) {
    console.error('Error fetching all users:', error)
    return {
      items: [],
      total: 0,
      page: 1,
      per_page: 20,
      total_pages: 0,
      has_next: false,
      has_prev: false,
    }
  }
}

// ============================================
// LIST ALL PROJECTS (Listar todos los proyectos)
// ============================================

export async function fetchAllTenants(params?: {
  page?: number
  per_page?: number
  search?: string
  status?: string
}): Promise<PaginatedResult<Project>> {
  try {
    const query = new URLSearchParams()
    if (params?.page) query.append('page', params.page.toString())
    if (params?.per_page) query.append('per_page', params.per_page.toString())
    if (params?.search) query.append('search', params.search)
    if (params?.status) query.append('status', params.status)
    
    const response = await api.get<ProjectsListResponse>(
      `/admin/projects${query.toString() ? '?' + query.toString() : ''}`
    )
    
    return {
      items: response.projects.map(mapProjectFromBackend),
      total: response.pagination.total,
      page: response.pagination.page,
      per_page: response.pagination.per_page,
      total_pages: response.pagination.total_pages,
      has_next: response.pagination.has_next,
      has_prev: response.pagination.has_prev,
    }
  } catch (error) {
    console.error('Error fetching all projects:', error)
    return {
      items: [],
      total: 0,
      page: 1,
      per_page: 20,
      total_pages: 0,
      has_next: false,
      has_prev: false,
    }
  }
}

// ============================================
// GET AUDIT LOGS (Obtener logs de auditoría)
// ============================================

export async function fetchAuditLogs(params?: {
  page?: number
  per_page?: number
  user_id?: string
  action?: string
  days?: number
}): Promise<PaginatedResult<AuditLog>> {
  try {
    const query = new URLSearchParams()
    if (params?.page) query.append('page', params.page.toString())
    if (params?.per_page) query.append('per_page', params.per_page.toString())
    if (params?.user_id) query.append('user_id', params.user_id)
    if (params?.action) query.append('action', params.action)
    if (params?.days) query.append('days', params.days.toString())
    
    const response = await api.get<AuditLogsResponse>(
      `/admin/audit-logs${query.toString() ? '?' + query.toString() : ''}`
    )
    
    return {
      items: response.logs.map(mapAuditLogFromBackend),
      total: response.pagination.total,
      page: response.pagination.page,
      per_page: response.pagination.per_page,
      total_pages: response.pagination.total_pages,
      has_next: response.pagination.has_next,
      has_prev: response.pagination.has_prev,
    }
  } catch (error) {
    console.error('Error fetching audit logs:', error)
    return {
      items: [],
      total: 0,
      page: 1,
      per_page: 20,
      total_pages: 0,
      has_next: false,
      has_prev: false,
    }
  }
}

// ============================================
// UPDATE USER STATUS (Activar/Desactivar usuario)
// ============================================

export async function toggleUserStatus(
  userId: string,
  status: 'active' | 'disabled'
): Promise<boolean> {
  try {
    await api.patch(`/admin/users/${userId}/status`, { status })
    return true
  } catch (error) {
    console.error('Error updating user status:', error)
    return false
  }
}

// ============================================
// UPDATE PROJECT STATUS (Activar/Desactivar proyecto)
// ============================================

export async function toggleTenantStatus(
  projectId: string,
  status: 'active' | 'disabled'
): Promise<boolean> {
  try {
    await api.patch(`/admin/projects/${projectId}/status`, { status })
    return true
  } catch (error) {
    console.error('Error updating project status:', error)
    return false
  }
}

// ============================================
// GET GLOBAL STATS (Obtener estadísticas globales)
// ============================================

export async function getGlobalStats(): Promise<GlobalStatsResponse | null> {
  try {
    const response = await api.get<GlobalStatsResponse>('/admin/stats')
    return response
  } catch (error) {
    console.error('Error getting global stats:', error)
    return null
  }
}

// ============================================
// HEALTH CHECK (Verificar salud del sistema)
// ============================================

export async function fetchHealthChecks(): Promise<HealthCheck[]> {
  try {
    const response = await api.get<HealthCheckResponse>('/admin/health')
    
    // Convertir la respuesta del backend al formato esperado por el frontend
    return [
      {
        service: 'API',
        status: response.status,
        latency: 0,
        last_check: response.timestamp,
      },
      {
        service: 'Database',
        status: response.database.status,
        latency: response.database.latency,
        last_check: response.timestamp,
      }
    ]
  } catch (error) {
    console.error('Error fetching health checks:', error)
    return [
      {
        service: 'API',
        status: 'down',
        latency: 0,
        last_check: new Date().toISOString(),
      },
      {
        service: 'Database',
        status: 'down',
        latency: 0,
        last_check: new Date().toISOString(),
      }
    ]
  }
}

// ============================================
// HELPERS
// ============================================

/**
 * Obtener usuarios activos
 */
export async function getActiveUsers(page = 1, per_page = 20): Promise<PaginatedResult<User>> {
  return fetchAllUsers({ page, per_page, status: 'active' })
}

/**
 * Obtener usuarios desactivados
 */
export async function getDisabledUsers(page = 1, per_page = 20): Promise<PaginatedResult<User>> {
  return fetchAllUsers({ page, per_page, status: 'disabled' })
}

/**
 * Obtener proyectos activos
 */
export async function getActiveProjects(page = 1, per_page = 20): Promise<PaginatedResult<Project>> {
  return fetchAllTenants({ page, per_page, status: 'active' })
}

/**
 * Obtener proyectos desactivados
 */
export async function getDisabledProjects(page = 1, per_page = 20): Promise<PaginatedResult<Project>> {
  return fetchAllTenants({ page, per_page, status: 'disabled' })
}
